### PR TITLE
Fix kubelet on Windows fails if a pod has SecurityContext with RunAsUser

### LINKED ISSUE #125012: kubelet on Windows fails if a pod has SecurityContext with RunAsUser.
### What happened?

Install a cluster with v1.29.4 with linux node as master and worker node as windows
kubeadm upgrade apply v1.30.0 fails with

`[ERROR CreateJob]: Job \"upgrade-health-check-4rxpv\" in the namespace \"kube-system\" did not complete in 15s: no condition of type Complete [preflight] If you know what you are doing, you can make a check non-fatal with `--ignore-preflight-errors`

Describe Pod events:

`mountvolume.setup failed for volume "kube-api-access-rncjh" : chown c:\var\lib\kubelet\pods\c5a3db55-4ee9-4b99-8fd6-17da5799f34a\volumes\kubernetes.io~projected\kube-api-access-rncjh\..2024_05_08_20_28_34.3908400707\token: not supported by windows`

It seems that previously the pod was Pending as well, but this was ignored, because the jobs was successfully deleted in defer and return value was overridden with nil.

```
defer func() {
    lastError = deleteHealthCheckJob(client, ns, jobName)
}()
```

https://github.com/kubernetes/kubernetes/blob/v1.29.1/cmd/kubeadm/app/phases/upgrade/health.go#L151

### What did you expect to happen?

Upgrade health check job expected to be scheduled only on Linux nodes using node selectors, rather than on Windows nodes.

### How can we reproduce it (as minimally and precisely as possible)?

See **What happened?**

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

kubeadm version (use kubeadm version): v1.30.0

Environment:

Kubernetes version (use kubectl version): v1.30.0
Cloud provider or hardware configuration:
OS (e.g. from /etc/os-release): Ubuntu 22.04.2 LTS, Windows Server 2022 Datacenter
Kernel (e.g. uname -a): 5.19.0-1025-aws
Container networking plugin (CNI): calico

</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by carlory**
#### What type of PR is this?

/kind bug

#### Which issue(s) this PR fixes:

Fixes #125012 

#### Does this PR introduce a user-facing change?

```release-note
Fix kubelet on Windows fails if a pod has SecurityContext with RunAsUser
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/volume/util/atomic_writer.go [hunk 1]
```diff
@@ -444,9 +444,12 @@ func (w *AtomicWriter) writePayloadToDir(payload map[string]FileProjection, dir
...
-		if err := os.Chown(fullPath, int(*fileProjection.FsUser), -1); err != nil {
-			klog.Errorf("%s: unable to change file %s with owner %v: %v", w.logContext, fullPath, int(*fileProjection.FsUser), err)
-			return err
+
+		if runtime.GOOS != "windows" {
+			if err := os.Chown(fullPath, int(*fileProjection.FsUser), -1); err != nil {
+				klog.Errorf("%s: unable to change file %s with owner %v: %v", w.logContext, fullPath, int(*fileProjection.FsUser), err)
+				return err
+			}
 		}
```
Comment (neolit123) [inline pkg/volume/util/atomic_writer.go:453]:
```suggestion
		if runtime.GOOS != "windows" {
			if err := os.Chown(fullPath, int(*fileProjection.FsUser), -1); err != nil {
				klog.Errorf("%s: unable to change file %s with owner %v: %v", w.logContext, fullPath, int(*fileProjection.FsUser), err)
				return err
			}
		} else {
			klog.Warningf("%s: skipping change of Linux owner %v for file %s; unsupported on Windows", w.logContext, int(*fileProjection.FsUser), fullPath)
		}
```

**Review by neolit123 — COMMENTED** [review]
istead of doing the if runtime.GOOS check it' s better to make it compile time.
add a new file called chown_linux.go and chown_unsupported.go

chown_unsupported.go will have the !linux build tag.

the function can be called Chown and it will accept the:
- logger context
- Linux user
- file path

it can return error on Linux and always nil on Windows

**Comment by neolit123** [comment]
windows e2e seems to have failed.

**Comment by carlory** [comment]
> windows e2e seems to have failed.

It's my mistaken. I should be fixed now.

**Comment by carlory** [comment]
> istead of doing the if runtime.GOOS check it' s better to make it compile time.

updated.

**Review by neolit123 — COMMENTED** [review]
thanks!
/lgtm

**Review by neolit123 — COMMENTED** [review]
/lgtm
/approve

thanks!

**Review by mrunalp — APPROVED** [review]

**Comment by jsafrane** [comment]
/approve

This makes RunAsUser consistent with [FSGroup](https://github.com/kubernetes/kubernetes/blob/58fe98e2cc7f8efac886765af230a48285d7b9b6/pkg/volume/volume_unsupported.go#L27) on Windows.

**Comment by carlory** [comment]
@jsafrane do we need to cherry-pick it into supported release branches?

**Pull request merged**